Decimal DIV changes

razajafri opened the pull request:
This PR deals with the changes to the Decimal `DIV` in cudf.

There is a bug in rapidsai/cudf#7435 where it's returning a Decimal col when a BOOL8 col is expected. I have commented on the PR https://github.com/rapidsai/cudf/pull/7435#issuecomment-792217382

Signed-off-by: Raza Jafri <rjafri@nvidia.com>

<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

razajafri commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala:
```diff
@@ -239,22 +239,99 @@ object GpuDivideUtil {
+    } else {
+      l.scale - outputScale
+    }
+    val newType = DecimalType(Math.min(18, l.precision + newScale), Math.min(18, newScale))
```
Good catch, there already is one that I can use Decimal.MAX_LONG_DIGITS()

razajafri commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuExpressions.scala:
```diff
@@ -231,12 +241,17 @@ trait CudfBinaryExpression extends GpuBinaryExpression {
+      BinaryOperable.implicitConversion(binaryOp, lBase, rhs)
+    } else {
+      outputType(lBase, rhs)
+    }
```
Thanks, i have updated the code

razajafri commented:
@gerashegalov have I addressed all your concerns? PTAL

razajafri commented:
@revans2 can you take a look at this PR.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala:
```diff
@@ -239,22 +239,98 @@ object GpuDivideUtil {
   }
 }
 
+abstract class GpuDecimalDivide(left: Expression, right: Expression) extends GpuDivModLike with
```
Why create an abstract class when only one thing implements it?

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala:
```diff
@@ -239,22 +239,98 @@ object GpuDivideUtil {
+    case _ => super.dataType
+  }
+
+  def getNewDecimalType(outputScale: Int, l: DecimalType, r: DecimalType): DecimalType = {
```
The name of this method does not explain anything about what it is doing, and there are no comments to explain it either.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala:
```diff
@@ -239,22 +239,98 @@ object GpuDivideUtil {
+  }
+
+  def getNewDecimalType(outputScale: Int, l: DecimalType, r: DecimalType): DecimalType = {
+    val newScale = if (outputScale + r.scale > l.scale) {
```
FYI: Spark has already make the lhs and rhs precision and scale the same.  This is part of the problem with multiply and divide is that Spark makes them both longer than they need to be so that they can be the same, which makes the spark binary op code happen, but results Spark doing things really inefficiently.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala:
```diff
@@ -239,22 +239,98 @@ object GpuDivideUtil {
+    } else {
+      l.scale - outputScale
+    }
+    val newType = DecimalType(Math.min(Decimal.MAX_LONG_DIGITS, l.precision + newScale),
```
I thought that the requirement was `r.scale + l.scale == outputScale`. This feels rather convoluted to get there.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala:
```diff
@@ -239,22 +239,98 @@ object GpuDivideUtil {
+    val newType = DecimalType(Math.min(Decimal.MAX_LONG_DIGITS, l.precision + newScale),
+      Math.min(Decimal.MAX_LONG_DIGITS, newScale))
+    val raise = newScale - l.scale
+    require(raise <= Decimal.MAX_LONG_DIGITS, s"$newType  is not supported for GPU processing yet.")
```
Is there ever a case where this will throw? If so then we really need the checks in tagging before we get to the point that this would throw an exception.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala:
```diff
@@ -239,22 +239,98 @@ object GpuDivideUtil {
+      case (l: DecimalType, r: DecimalType) => {
+        val outputScale = dataType.asInstanceOf[DecimalType].scale
+        val newType = getNewDecimalType(outputScale, l, r)
+        if (outputScale + r.scale > l.scale) {
```
Again r.scale == l.scale in all cases for Spark right now.

razajafri commented:
@revans2 I have addressed your concerns. Knowing what I know now, the math is pretty much just adding the r.scale to the outputType.scale so after div is called we end up with outputType.scale.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala:
```diff
@@ -245,16 +245,58 @@ case class GpuDivide(left: Expression, right: Expression) extends GpuDivModLike
+    (left.dataType, right.dataType) match {
+      case (_: DecimalType, r: DecimalType) => {
+        val outType = dataType.asInstanceOf[DecimalType]
+        val newType = DecimalType(outType.precision, outType.scale + r.scale)
```
Can you please add a comment to explain the rules here about what is needed for the divide to work? It looks like from the CUDF code that for a divide the output scale must be `lhs.scale - rhs.scale = outputScale`.  I did a much of math and I think casting the lhs up is going to be fine until we hit 128-bit support for decimal and need to potentially check for overflow, so it would be good to have a comment about that here too.

razajafri replied:
OK, I will do that. I thought it was unnecessary as we are already checking for it in `tagExprForGpu`

razajafri replied:
nevermind, we are only checking for the intermediary result. 

I think we should add a check for `GpuDivideUtil.decimalDataType + r.scale` instead of what we are doing now? What do you think?

razajafri commented:
@revans2 can you check my math on the `tagExprForGpu` method?

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:
```diff
@@ -1733,7 +1733,25 @@ object GpuOverrides {
             case (l: DecimalType, r: DecimalType) =>
-              val intermediateResult = GpuDivideUtil.decimalDataType(l, r)
+              val outputType = GpuDivideUtil.decimalDataType(l, r)
+              // We will never hit a case where outputType.precision < outputType.scale + r.scale.
```
I still want a comment about the 128 bit use case

razajafri replied:
Thanks for the clarification, I have added the

pxLi commented:
FYI the new cudf has just been built and deployed to snapshot

jlowe commented:
build

jlowe commented:
build

jlowe reviewed (approved):
Premerge is failing for 3.1.2 tests because a recent 3.2.0 fix was cherry-picked.  All other tests passed.  Committing this since it is blocking all premerge CI for other PRs, including the one that will fix the 3.1.2 tests.
